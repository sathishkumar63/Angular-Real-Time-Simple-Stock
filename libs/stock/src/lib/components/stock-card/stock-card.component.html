<div class="stock-card__group">
  <div
    class="stock-card__item"
    *ngFor="let stock of stocksList"
    [class.stock-card__item--active]="stock.isActive"
  >
    <p class="stock-card__item__title">
      {{ stock.symbol | uppercase }}
    </p>
    <p class="stock-card__item__current-price f-20">
      {{ stock.currentPrice | currency: stock.currency }}
    </p>
    <p class="stock-card__item__daily-high-price f-20">
      {{ stock.dayHighPrice | number: '1.2-2' }}
    </p>
    <p class="stock-card__item__daily-low-price f-20">
      {{ stock.dayLowPrice | number: '1.2-2' }}
    </p>
    <ng-container *ngIf="!(devicesService.isMobile$ | async)">
      <p class="stock-card__item__fiftyTwoWeekLow f-20">
        {{ stock.fiftyTwoWeekLow | number: '1.2-2' }}
      </p>
      <p class="stock-card__item__fiftyTwoWeekHigh f-20">
        {{ stock.fiftyTwoWeekHigh | number: '1.2-2' }}
      </p>
      <p class="stock-card__item__regularMarketVolume f-20">
        {{ stock.regularMarketVolume | number }}
      </p>
    </ng-container>
    <label class="stock-card__item__switch">
      <input
        type="checkbox"
        class="stock-card__item__checkbox"
        [checked]="stock.isActive"
        (click)="onToggle(stock, $event.target.checked)"
      />
      <span class="stock-card__item__switch--slider"></span>
    </label>
  </div>
</div>
