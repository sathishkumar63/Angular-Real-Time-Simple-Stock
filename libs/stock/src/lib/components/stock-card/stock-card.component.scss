.stock-card {
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 2.5rem 4rem 1rem;
  margin: 1rem;
  background-color: #4caf50;
  color: #263238;
  border-radius: 10px;
  &__title {
    font-weight: 700;
    font-size: 36px;
    margin: 0 0 1rem;
  }
  &__current-price {
    font-size: 24px;
    margin: 0 0 1rem;
  }
  &__daily-high-price,
  &__daily-low-price {
    font-size: 20px;
    margin: 0;
  }
  &__vol {
    font-size: 26px;
    margin: 1rem 0;
  }
  &__switch {
    position: relative;
    display: inline-block;
    width: 100px;
    height: 40px;
    input {
      width: 0;
    }
    &--slider:after {
      content: 'OFF';
      display: block;
      position: absolute;
      transform: translate(-50%, -50%);
      top: 50%;
      left: 60%;
      font-size: 1rem;
      font-weight: 700;
      color: #ffffff;
    }
    input:checked + .stock-card__switch--slider {
      background-color: #00476a;
      &:before {
        transform: translateX(60px);
        background-color: #009688;
      }
      &:after {
        content: 'ON';
        left: 40%;
        color: #009688;
      }
    }
    &--slider {
      position: absolute;
      cursor: pointer;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      background-color: #bfbfbf;
      transition: 0.4s;
      border-radius: 40px;
      &:before {
        position: absolute;
        content: '';
        height: 30px;
        width: 30px;
        left: 4px;
        bottom: 5px;
        background-color: #6e7781;
        transition: 0.4s;
        border-radius: 50%;
      }
    }
  }
}
