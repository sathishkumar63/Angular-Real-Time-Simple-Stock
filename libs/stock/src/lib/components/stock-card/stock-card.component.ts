import {
  Component,
  EventEmitter,
  Input,
  Output,
  ViewEncapsulation,
} from '@angular/core';

@Component({
  selector: 'stock-card',
  templateUrl: './stock-card.component.html',
  styleUrls: ['./stock-card.component.scss'],
  encapsulation: ViewEncapsulation.None,
})
export class StockCardComponent {
  @Input() stocksList: any[];
  @Output() notifyToggle = new EventEmitter();

  onToggle(stock: any, event: any) {
    setTimeout(
      () =>
        this.notifyToggle.emit({
          ...stock,
          isActive: event.target.checked,
        }),
      300
    );
  }
}
