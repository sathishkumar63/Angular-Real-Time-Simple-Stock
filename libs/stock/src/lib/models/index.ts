export * from './stock.constant';
export * from './stock.interface';
export * from './stock.mock.data';
