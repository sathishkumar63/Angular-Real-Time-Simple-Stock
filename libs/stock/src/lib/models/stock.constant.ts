export const stocksSymbol: string[] = [
  'GOOG',
  'GOOGL',
  'MSFT',
  'AAPL',
  'TSLA',
  'AMZN',
  'UBER',
  'DIDI',
  'META',
  'NFLX',
  'CHIK',
  'GLD',
  'IAU',
  'BTC-USD',
  'ETH-USD',
  'DOGE-USD',
  'SHIB-USD',
  'ETH-CAD',
  'USDT-USD',
  'USDC-USD',
];
