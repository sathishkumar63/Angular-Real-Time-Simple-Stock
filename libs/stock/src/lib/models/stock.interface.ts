export interface IStock {
  name?: string;
  symbol: string;
  currency?: string;
  currentPrice: number;
  dayHighPrice?: number;
  dayLowPrice?: number;
  fiftyTwoWeekLow?: number;
  fiftyTwoWeekHigh?: number;
  regularMarketVolume?: number;
  isActive?: boolean;
}
