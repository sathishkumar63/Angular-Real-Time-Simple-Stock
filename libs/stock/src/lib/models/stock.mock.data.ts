import { StocksFeatureState, StocksState } from '../+state';
import { IStock } from './stock.interface';

export const mockStocks: IStock[] = [
  {
    name: 'Alphabet',
    symbol: 'GOOG',
    currency: 'USD',
    currentPrice: 2228.55,
    dayHighPrice: 2269.94,
    dayLowPrice: 2217.22,
    fiftyTwoWeekLow: 2044.16,
    fiftyTwoWeekHigh: 3042,
    regularMarketVolume: 1567487,
  },
  {
    name: 'Microsoft',
    symbol: 'MSFT',
    currency: 'USD',
    currentPrice: 252.99,
    dayHighPrice: 260.54,
    dayLowPrice: 252.86,
    fiftyTwoWeekLow: 246.44,
    fiftyTwoWeekHigh: 349.67,
    regularMarketVolume: 31445841,
  },
];

export const mockStocksState: StocksState = {
  data: mockStocks,
  loading: false,
  loaded: true
};

export const mockStocksFeatureState: StocksFeatureState = {
  stocks: mockStocksState
};
