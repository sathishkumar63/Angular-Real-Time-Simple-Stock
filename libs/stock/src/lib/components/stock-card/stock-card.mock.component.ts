import { Component, Input } from '@angular/core';
import { IStock } from '../../models';

@Component({
  selector: 'stock-card',
  template: ''
})
export class MockStockCardComponent {
  @Input() stocksList: IStock[];
}
