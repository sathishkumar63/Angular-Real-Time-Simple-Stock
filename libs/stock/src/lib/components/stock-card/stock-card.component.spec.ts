import {
  ComponentFixture,
  fakeAsync,
  TestBed,
  tick,
} from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { mockStocks } from '../../models';
import { StockCardComponent } from './stock-card.component';

describe('StockCardComponent', () => {
  let component: StockCardComponent;
  let fixture: ComponentFixture<StockCardComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [StockCardComponent],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(StockCardComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render stocks card when the stocksList.length > 0', () => {
    // arrange
    component.stocksList = mockStocks;
    // act
    fixture.detectChanges();
    const stockCardGroup = fixture.debugElement.query(
      By.css('.stock-card__group')
    ).nativeElement;
    // assert
    expect(stockCardGroup).toMatchSnapshot();
  });

  it('onToggle should be called when clicking the toggle', () => {
    // arrange
    jest.spyOn(component, 'onToggle');
    component.stocksList = mockStocks;
    // act
    fixture.detectChanges();
    const toggle = fixture.debugElement.queryAll(
      By.css('.stock-card__item__checkbox')
    )[0];
    toggle.nativeElement.click();
    fixture.detectChanges();
    // assert
    expect(component.onToggle).toBeCalledWith(mockStocks[0], true);
  });

  it('should emit notifyToggle event when onToggle method invoked', fakeAsync(() => {
    // arrange
    jest.spyOn(component.notifyToggle, 'emit');
    const isChecked = true;
    // act
    component.onToggle(mockStocks[0], isChecked);
    tick(300);
    // assert
    expect(component.notifyToggle.emit).toBeCalledWith({
      ...mockStocks[0],
      isActive: isChecked,
    });
  }));
});
