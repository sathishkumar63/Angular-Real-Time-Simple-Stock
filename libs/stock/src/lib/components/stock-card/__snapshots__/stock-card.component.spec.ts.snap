// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockCardComponent should render stocks card when the stocksList.length > 0 1`] = `
<div
  class="stock-card__group"
>
  
  <div
    class="stock-card__item"
  >
    <p
      class="stock-card__item__title"
    >
       GOOG 
    </p>
    <p
      class="stock-card__item__current-price f-20"
    >
       $2,228.55 
    </p>
    <p
      class="stock-card__item__daily-high-price f-20"
    >
       2,269.94 
    </p>
    <p
      class="stock-card__item__daily-low-price f-20"
    >
       2,217.22 
    </p>
    
    
    <p
      class="stock-card__item__fiftyTwoWeekLow f-20"
    >
       2,044.16 
    </p>
    <p
      class="stock-card__item__fiftyTwoWeekHigh f-20"
    >
       3,042.00 
    </p>
    <p
      class="stock-card__item__regularMarketVolume f-20"
    >
       1,567,487 
    </p>
    <label
      class="stock-card__item__switch"
    >
      <input
        class="stock-card__item__checkbox"
        type="checkbox"
      />
      <span
        class="stock-card__item__switch--slider"
      />
    </label>
  </div>
  <div
    class="stock-card__item"
  >
    <p
      class="stock-card__item__title"
    >
       MSFT 
    </p>
    <p
      class="stock-card__item__current-price f-20"
    >
       $252.99 
    </p>
    <p
      class="stock-card__item__daily-high-price f-20"
    >
       260.54 
    </p>
    <p
      class="stock-card__item__daily-low-price f-20"
    >
       252.86 
    </p>
    
    
    <p
      class="stock-card__item__fiftyTwoWeekLow f-20"
    >
       246.44 
    </p>
    <p
      class="stock-card__item__fiftyTwoWeekHigh f-20"
    >
       349.67 
    </p>
    <p
      class="stock-card__item__regularMarketVolume f-20"
    >
       31,445,841 
    </p>
    <label
      class="stock-card__item__switch"
    >
      <input
        class="stock-card__item__checkbox"
        type="checkbox"
      />
      <span
        class="stock-card__item__switch--slider"
      />
    </label>
  </div>
</div>
`;
